### DISCUSSION TITLE
Basic question about variables...

### DISCUSSION THREAD

**Discussion by lurendrejer** [zwave-js/node-red-contrib-zwave-js#123] [Q&A]
Im fiddling with a 'version 2' of my z-wave command subflow...
I've created a few variables but how on earth do i get these inside an object?

![image](https://user-images.githubusercontent.com/30922656/136080416-edbedf4d-656d-43f5-8f12-8eb5516ab5e7.png)

I have tried quite a few things from the node-red documentation, without much luck...
I know i'll have to translate 'on' to a Bool, etc. - but first I have to find out how to use the variables.

//argh-test

let Message = {
    payload: {
        mode: "CCAPI",
        cc: "$commandtype",
        method: "set",
        params: $env('state'),
        endpoint: "${endpoint}"
    }
}
return Message

**Comment by lurendrejer** [comment]
Did this instead.. I don't know if it impacts performance or if it just (more or less) gets converted to the above, when used by NR.

![image](https://user-images.githubusercontent.com/30922656/136084327-8d81d4f3-3b8b-4df0-b95e-6c75edd13bb1.png)

![image](https://user-images.githubusercontent.com/30922656/136090102-92ca8818-04c0-4212-ba84-a259b8ef4839.png)

**Comment by marcus-j-davies** [comment]
Hey,

In a JS function node you can get the variables declared in your subflow as follow.

```javascript
let CommandClass = env.get("commandclass");
let Method = env.get("classMethod");
let Value = env.get("state");
let EP = parseInt(env.get("endpoint"));
let Node = parseInt(env.get("node"));

let Message = {
    payload: {
        mode: "CCAPI",
        node: Node,
        cc: CommandClass,
        method: Method,
        params: [Value], /* remember to keep params as an array */
        endpoint: EP
    }
}
return Message
```

We have an awesome node being introduced soon (Known as the **cmd-factory**) - that will generate commands for you (given a set of inputs).

![image](https://user-images.githubusercontent.com/55892693/136086064-f9489d88-63f8-43be-9edf-accadb2a93ad.png)

**Reply by lurendrejer** [reply]
yup, i know - but i was already creating a setpoint-thermostat selection thingy - and wanted to tinker :)

![image](https://user-images.githubusercontent.com/30922656/136089450-d7e1200d-f2c9-4c93-8d7f-214f12c56153.png)

**Reply by marcus-j-davies** [reply]
it wont take you long until realise you have an addiction problem 🤣 

![image](https://user-images.githubusercontent.com/55892693/136089835-d7233db6-5993-48ae-b71d-f31d13c8f505.png)

**Comment by marcus-j-davies** [comment]
Oh, and in answer to you performance worry.
node-red is ridiculously optimised  - I wouldn't worry.

My entire home security system - is run by, managed by and monitored by node red with various motion sensors, door sensors, keypad door entry devices etc etc - all runs without a hitch.

**Reply by lurendrejer** [reply]
I knew it would be a minimal performance penalty, if any.
But i guess the gui-draggy-droppe stuff gets converted to js in some way, in the end.

**Reply by lurendrejer** [reply]
And, yes - NR is stupidly fast...
